### ISSUE TITLE
[REVIEW]: MyQueue: Task and workflow scheduling system

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1844]
**Submitting author:** @jjmortensen (<a href="http://orcid.org/0000-0001-5090-6706">Jens Jørgen Mortensen</a>)
**Repository:** <a href="https://gitlab.com/myqueue/myqueue" target ="_blank">https://gitlab.com/myqueue/myqueue</a>
**Version:** version-20.1.1
**Editor:** @danielskatz
**Reviewer:** @gonsie, @marksantcroos
**Archive:** <a href="https://doi.org/10.5281/zenodo.3607221" target="_blank">10.5281/zenodo.3607221</a>

## Status

[![status](https://joss.theoj.org/papers/33edc00feaacaaaca2bb0e22011bb421/status.svg)](https://joss.theoj.org/papers/33edc00feaacaaaca2bb0e22011bb421)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/33edc00feaacaaaca2bb0e22011bb421"><img src="https://joss.theoj.org/papers/33edc00feaacaaaca2bb0e22011bb421/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/33edc00feaacaaaca2bb0e22011bb421/status.svg)](https://joss.theoj.org/papers/33edc00feaacaaaca2bb0e22011bb421)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://gitlab.com/myqueue/myqueue" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@gonsie & @marksantcroos, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @danielskatz know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://gitlab.com/myqueue/myqueue">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@jjmortensen) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://gitlab.com/myqueue/myqueue">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@jjmortensen) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @gonsie, @marksantcroos it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01844/joss.01844/10.21105.joss.01844.pdf)

**Comment by danielskatz** [comment]
👋 @gonsie, @marksantcroos - as I think you know, we'll do the review here.  Your job is to work through the review checklists, checking off items or creating new issues in the target repository and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

If you have any questions or problems, please let me know

**Comment by gonsie** [comment]
Thanks for submission @jjmortensen. I've created 2 issues in the repo:
- [#9 Move license text to LICENSE](https://gitlab.com/myqueue/myqueue/issues/9)
- [#10 Assert error from `mq test`](https://gitlab.com/myqueue/myqueue/issues/10)
- [#11 Ability to kill the daemon?](https://gitlab.com/myqueue/myqueue/issues/11)

For the paper, could you include some details on how MyQueue compares with other tools?  I think the most relevant tools would come from the workflow tools / managers.

**Comment by danielskatz** [comment]
👋 @marksantcroos - how are you doing on your review?

**Comment by danielskatz** [comment]
@jjmortensen - fwiw, a long list of workflow tools is in https://github.com/common-workflow-language/common-workflow-language/wiki/Existing-Workflow-systems (and a shorter list is in https://github.com/meirwah/awesome-workflow-engines)

**Comment by jjmortensen** [comment]
Thanks for the links.  I'm familiar with Fireworks and Aiida.  Will take a look ...

**Comment by marksantcroos** [comment]
I've also started my review by going through installation and testing, but got stuck rather quickly.

- [#13 system wide installation vs virtual environment](https://gitlab.com/myqueue/myqueue/issues/13)
- [#14 echo command to create hello.py is wrong in tutorial](https://gitlab.com/myqueue/myqueue/issues/14)
- [#15 sbatch: unrecognized option '--workdir'](https://gitlab.com/myqueue/myqueue/issues/15)

Although I might be biased, these errors do trigger my "yet another workflow system that tries (and fails) to integrate with queuing systems".

Putting this out here for now, will continue the review and edit accordingly.

**Comment by jjmortensen** [comment]
Thanks for all your comments so far.  There is now a version 19.11.1 available on PyPI with some new features and bug-fixes.  There is also an updated version of the paper.  See:

https://myqueue.readthedocs.io/en/latest/releasenotes.html
https://gitlab.com/myqueue/myqueue/merge_requests/40

**Comment by danielskatz** [comment]
When you update the paper, please rebuild it here, as I am now going to do

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01844/joss.01844/10.21105.joss.01844.pdf)

**Comment by gonsie** [comment]
This paper is still missing a ‘state of the field’ summary. Comparing MQ to Slurm/PBS is not the right level of comparison. It needs to be compared to other workflow tools. 

In addition, the paper mentions that MQ was used to drive a study of 10,000 jobs. Please provide details on that. How did it handle this large number of jobs? What was the job limit enforced by the underlying scheduler? I believe that this may be the real contribution of MQ (rather than just a front end to a scheduler).

**Comment by danielskatz** [comment]
👋 @jjmortensen - note that there are some questions/actions for you here.

**Comment by danielskatz** [comment]
👋 @marksantcroos - Can you let us know how your review is proceeding? Are you stuck on anything?

**Comment by jjmortensen** [comment]
We are looking into it ... in the meantime, @gonsie I added a small section on the web-page describing how to work with many jobs:

https://myqueue.readthedocs.io/en/latest/workflows.html#handling-very-many-tasks

**Comment by mortengjerding** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01844/joss.01844/10.21105.joss.01844.pdf)

**Comment by mortengjerding** [comment]
We have now added a section describing the state of the field, where we compare MyQueue to AiiDA and Fireworks, trying to emphasize what MyQueue can/cannot do. Let us know if you have suggestions for further improvements.

**Comment by marksantcroos** [comment]
I've worked some more with the software and got something out of it. Will go over the new pdf and report my findings.

**Comment by danielskatz** [comment]
@marksantcroos - thanks

**Comment by danielskatz** [comment]
@marksantcroos  - any update?

**Comment by marksantcroos** [comment]
@jjmortensen @mortengjerding Thanks for the tool. The principle of having a "local" and optionally per directory queue is an interesting concept. I tried it out for some data crunching I had to do and ultimately managed to get the job done, which is some sign of success :-)

The paper positions adequately (albeit minimally) compared to other tools out there.

I've opened up another issue in the code repo, but that shouldn't block the paper.

@danielskatz I've checked all boxes and accept it as is now.

**Comment by danielskatz** [comment]
Thanks @marksantcroos

**Comment by danielskatz** [comment]
@gonsie - can I check in with you to get a status of where things are and what you think needs to be done before you can check off your remaining boxes?

**Comment by gonsie** [comment]
I'm still having trouble actually using myqueue. I just filed another issue in the repo:

- [#17 Config file documentation is incomplete](https://gitlab.com/myqueue/myqueue/issues/17)

**Comment by gonsie** [comment]
In response to the discussion about submitting a large number of tasks...

> We are looking into it ... in the meantime, @gonsie I added a small section on the web-page describing how to work with many jobs:
> 
> https://myqueue.readthedocs.io/en/latest/workflows.html#handling-very-many-tasks

MyQueue should probably have a sensible default. Unlimited is not a good default. Maybe 1000?

**Comment by gonsie** [comment]
A contributing guide would be nice, but I guess it's not required.

**Comment by danielskatz** [comment]
Actually, some guidance for contributors is required, as stated in the **Community guidelines** review criterion

> Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

**Comment by danielskatz** [comment]
The README needs a fair amount more content, probably take from what's in the docs. Some of the following should be copied:

- A statement of need - What problems is the software designed to solve and who is the target audience
- Installation instructions, including a clearly-stated list of dependencies
- Example usage - or a pointer into the docs that has this
- Functionality documentation - or a pointer into the docs that has this
- Automated tests - or a pointer into the docs that has this
- Community guidelines - clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

Basically, the README should give enough info that someone can understand what the software is, how to build it, and what to do if they have problems.

**Comment by jjmortensen** [comment]
There is a new README now: https://gitlab.com/myqueue/myqueue/

**Comment by danielskatz** [comment]
That looks good to me, but I also want to hear @gonsie's opinion

**Comment by gonsie** [comment]
Readme looks good. I'm still working in issue #17 to get mq to actually work for me. 

I'll be gone until early January, happy holidays everyone!

**Comment by danielskatz** [comment]
Thanks - have a good break!

**Comment by danielskatz** [comment]
👋 @gonsie - just a reminder on this once you are back...

**Comment by gonsie** [comment]
The updated readme looks really nice and I've been able to successfully traverse the 'quick start' guide. I'll try to poke around a bit more this week to further test the functionality... but my review is close to being done 👍

**Comment by gonsie** [comment]
My review is complete. Looks great to me!

**Comment by danielskatz** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1088/2053-1583/aacfc1 is OK
- 10.26434/chemrxiv.9971054.v1 is OK
- 10.1002/cpe.3505 is OK
- https://doi.org/10.1016/j.commatsci.2015.09.013 is OK

MISSING DOIs

- https://doi.org/10.1007/10968987_3 may be missing for title: SLURM: Simple Linux Utility for Resource Management

INVALID DOIs

- None
```

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01844/joss.01844/10.21105.joss.01844.pdf)

**Comment by danielskatz** [comment]
@jjmortensen - it looks like one of your references needs a DOI to be added
Also, please use {}s in the tex to protect the cases (upper, lower) where needed - for example, JSSPP in that reference.  Once you've made changed, you can add a new comment here with ```@whedon generate pdf``` to generate a new version.  I will also proof-read the paper fairly soon.

**Comment by jjmortensen** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01844/joss.01844/10.21105.joss.01844.pdf)

**Comment by jjmortensen** [comment]
Added the missing DOI and upper cased JSSPP

**Comment by danielskatz** [comment]
@whedon accept

**Comment by whedon** [comment]
No archive DOI set. Exiting...

**Comment by danielskatz** [comment]
At this point could you:
- [ ] Make a tagged release of your software, and list the version tag of the archived version here.
- [ ] Archive the reviewed software in Zenodo or a similar service (e.g. figshare, an institutional repository)
- [ ] Check the archival deposit (e.g., in Zenodo) has the correct metadata, this includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it); you may also add the authors' ORCID.
- [ ] Please list the DOI of the archived version here.

I can then move forward with accepting the submission.

**Comment by jjmortensen** [comment]
The MyQueue version is 20.1.1 and the tag name in the Git repository is "version-20.1.1":

    https://gitlab.com/myqueue/myqueue/-/tags/version-20.1.1

I've uploaded an archive to Zenodo:

    https://zenodo.org/record/3607221

The DOI is "10.5281/zenodo.3607221".

**Comment by danielskatz** [comment]
@whedon set version-20.1.1 as version

**Comment by whedon** [comment]
OK. version-20.1.1 is the version.

**Comment by danielskatz** [comment]
@whedon set 10.5281/zenodo.3607221 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3607221" target="_blank">10.5281/zenodo.3607221</a> is the archive.

**Comment by danielskatz** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1088/2053-1583/aacfc1 is OK
- 10.1007/10968987_3 is OK
- 10.26434/chemrxiv.9971054.v1 is OK
- 10.1002/cpe.3505 is OK
- https://doi.org/10.1016/j.commatsci.2015.09.013 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1214

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1214, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by danielskatz** [comment]
👋 @jjmortensen  - please merge https://gitlab.com/myqueue/myqueue/merge_requests/51

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01844/joss.01844/10.21105.joss.01844.pdf)

**Comment by danielskatz** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1088/2053-1583/aacfc1 is OK
- 10.1007/10968987_3 is OK
- 10.26434/chemrxiv.9971054.v1 is OK
- 10.1002/cpe.3505 is OK
- https://doi.org/10.1016/j.commatsci.2015.09.013 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1215

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1215, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by danielskatz** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1217084335322533889) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1216
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01844](https://doi.org/10.21105/joss.01844)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by danielskatz** [comment]
Thanks to @gonsie, @marksantcroos for reviewing!
And congratulations to @jjmortensen and coauthors!

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.01844/status.svg)](https://doi.org/10.21105/joss.01844)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01844">
  <img src="https://joss.theoj.org/papers/10.21105/joss.01844/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.01844/status.svg
   :target: https://doi.org/10.21105/joss.01844
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.01844/status.svg)](https://doi.org/10.21105/joss.01844)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by jjmortensen** [comment]
Thanks a lot @danielskatz, @gonsie, @marksantcroos
